rfc/2024-11-13-provider-oci-registry.md

```diff
@@ -0,0 +1,240 @@
+
+---
+
+## User Documentation
+
+### For provider authors / mirror operators
+
+There are two possible approaches for distributing providers.
+
+<details>
+<summary>
+
+#### 📂 Standard container tooling approach
+
+</summary>
+
+As a provider author, you may want to package up your provider binary into an OCI image. This involves the following steps:
+
+1. Take all build artifacts and add them to a container image.
+2. Upload the container image to an OCI registry.
+
+For example, assuming you have your binary built and wanted to package it up into a container, you could use the following `Dockerfile`:
+
+```Dockerfile
+FROM scratch
+
+ADD terraform-provider-aws
+ADD README.md
+ADD LICENSE
+ADD CHANGELOG.md
+```
+
+You could now run `docker build -t ghcr.io/opentofu/terraform-provider-aws:1.2.3-amd64 .` followed by `docker push ghcr.io/opentofu/terraform-provider-aws:1.2.3`.
+
+For more advanced use cases, you could enable container image building in your [goreleaser config](https://goreleaser.com/customization/docker/) to enable building multi-arch images.
+</details>
+
+<details>
+<summary>
+
+#### 📂 ORAS approach
+
+</summary>
+
+As a provider author, you may want to package up your provider binary into an OCI image. You can do this by using the [ORAS command line tool](https://oras.land/docs/installation) using the following commands:
+
+```
+oras login -u username -p password ghcr.io
+oras push ghcr.io/opentofu/terraform-provider-aws:1.2.3 ./dist:application/octet-stream+tar
+```
+
+🛑🛑🛑 TODO: how do you create multi-arch with ORAS? 🛑🛑🛑
+
+</details>
+
+### Mirror configuration
+
+If you would like to use an OCI registry as your OpenTofu mirror, you will be able to do so by specifying a `provider_installation` block with an `oci_mirror` section as follows:
+
+```hcl
+provider_installation {
+  oci_mirror {
+    url = "ghcr.io/{{ .Namespace }}/terraform-provider-{{ .Name }}"
+  }
+}
+```
+
+Note that the `url` field is a Go template to resolve the container image address containing the provider binary. You can use `{{ .Namespace }}` to add the provider namespace (e.g. `hashicorp`) and `{{ .Name }}` to add the provider name (e.g. `aws`) to the container image address.
+
+> [!NOTE]
+> OpenTofu will **not** start a container from this image, but instead use the container image as a way to extract the provider binary. Also note that providers are not automatically available in any particular container registry. This feature is intended for advanced use cases.
+
+### The first-class configuration
+
+If you would like to use a provider hosted in an OCI registry, you may do so by invoking a well-known public registry in your `required_providers` block.
+
+```hcl
+terraform {
+  required_providers {
+    aws = {
+      source  = "ghcr.io/opentofu/terraform-provider-aws"
```

Comment:
In the first versions of this feature, we should make sure users are explicitly opt-in for using an OCI registry.

The previous solution with the tofurc file makes more sense from this point of view.

Reply:
I don't see why not using a oci://{host} URI format instead, so there's no need to special case some URIs

Reply:
I agree using the same prefix(`oci://`) as modules makes the most sense and solves the problem of whitelisting known hosts, which i feel takes some of the draw away from the feature.

Reply:
Unfortunately this discussion got split because we were covering some very similar ground in the feature request for supporting provider installation directly from GitHub. In this comment I explained why OpenTofu's provider source addresses are not URLs: https://github.com/opentofu/opentofu/issues/2109#issuecomment-2444518513

That comment describes it in more detail, but the short version is: modules are allowed to specify what providers they need using our abstract provider address syntax, but they are _not_ allowed to specify how those providers get installed because [provider installation strategies are customizable](https://opentofu.org/docs/cli/config/config-file/#provider-installation).

Accessing the registry whose hostname is included in the provider address is the default installation strategy for ease of use in simple cases, but lots of OpenTofu users run it in situations where direct access to the origin registry is not appropriate. The same would likely be true for direct access to arbitrary OCI registries, so for now we intend to continue using our abstract address syntax even when installing from OCI so that OCI can be the _default_ installation strategy while still allowing all of the alternative installation methods to work.

It remains to be seen (as we're discussing here) to what extent we will enable totally automatic installation from "well-known" OCI registries. Amongst other concerns, we need to figure out what (if anything) is the analog to GPG signing so that provider authors can specify which package checksums are "official".